libbpf-tools/readahead.c

```diff
@@ -119,7 +119,7 @@ int main(int argc, char **argv)
 		fprintf(stderr, "failed to open BPF object\n");
 		return 1;
 	}
-
+#ifndef __aarch64__
```

Comment:
the problem here is that you still have to handle function rename between kernel versions, just not for fentry but for kprobe. As is, tool will only work for v5.10+.

I think a bit cleaner solution would be to have both kprobe and fentry functions defined in BPF (with no arm64-specific `#ifndef` in BPF code). Then in user-space, do detection of all the cases we need to handle: arm64 vs non-arm64, do_page_cache_ra vs __do_page_cache_readahead.

To make this simpler, instead of using skeleton's readahead_bpf__attach(), we can attach each program individually. So in some cases you'll do bpf_program__attach_trace() (for fentry, preceded by correct bpf_program__set_attach_target), in others you'll do bpf_program__attach_kprobe() with the right kernel function name.

While at it, and with that approach, we can also add fallback to kprobe/kretprobe for x86, if we detect that fentry isn't supported. It's a bit complicated logic on user-space side, but we are trying to handle vastly differing environments, so I think it's justified.

WDYT?

Reply:
reworked; basic idea was to rely on the arch_prepare_bpf_trampoline return value which is (-ENOTSUPP) for everything except x86